This PR allows the use of (x, y) of the node as the (0, 0) position o…

cristianoccazinsp opened the pull request:
This PR allows the use of (x, y) of the node as the (0, 0) position of the image instead of using width/2 and height/2.

A new shape property was added: `useImageCoords` (default false).  This should not be a breaking change since it is false by default.

Closes https://github.com/visjs/vis-network/issues/811

TODO: Label positioning may not work as expected when this setting is used.

Thomaash commented on docs-kr/network/nodes.html:
```diff
@@ -206,6 +206,7 @@ <h3>옵션</h3>
       interpolation: false,  // only for image and circularImage shapes
       useImageSize: false,  // only for image and circularImage shapes
       useBorderWithImage: false  // only for image shape
+      useImageCoords: false  / only for image shape
```
syntax error

cristianoccazinsp replied:
Strange, I'm sure tests ran fine.

I do get the following file-related errors though:
<img width="753" alt="Screen Shot 2020-06-18 at 16 29 49" src="https://user-images.githubusercontent.com/48869228/85063812-10dc6c00-b181-11ea-9d7c-2eca56768987.png">

Thomaash replied:
`npx npm-run-all clean build test`

Thomaash commented on docs-kr/network/nodes.html:
```diff
@@ -1029,6 +1030,15 @@ <h3>옵션</h3>
+            <td class="indent">shapeProperties.useImageCoords</td>
+            <td>Boolean</td>
+            <td><code>false</code></td>
+            <td>This property only applies to the <code>image</code> shape.
+                When true, the image's (0,0) coords are used for drawing.
+                Note that label positioning may break with this option.
```
This should be Korean.

cristianoccazinsp replied:
I will do my best with google translate.

Thomaash commented on lib/network/modules/components/nodes/shapes/Image.js:
```diff
@@ -58,8 +58,15 @@ class Image extends CircleImageBase {
-    this.left = x - this.width / 2;
-    this.top = y - this.height / 2;
+
+    if(this.options.shapeProperties.useImageCoords){
+      this.left = x;
+      this.top = y;
+    }
+    else{
+      this.left = x - this.width / 2;
+      this.top = y - this.height / 2;
+    }
```
Could you format this better, please? Ideally like [this](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEACdqCWAzAFDAC0wGcA6CABxk2jOIIEMK4AFAJ0rjerjIFdicAJIBbBgHM4AYQgQ2AE2IBKYAB0oGDIRKkANnGwxUAXlQAPANzrN6bWRiUTqAJ5WNGAL7WMcXYLXumnZ6BkamZqgAtKjBAO6Y8oSoAPSoAExuNjFE9o6mzlHZOgRwmOIERqkZ3ugeIAA0IJTUtMigDGwcsSwdCMTIIAy6sQzO-Y0ARmwMYADWcDAAyhQzmFDiyDBsfHCNFSK6AOpE8MQrYHCLfZjUAG43zgNgxOMga4Lc7BJiyNhDgo0AFbEMwAIWmcwWiwYIjgABk1nBfv9diBgWZFmtxPoAIp8CDwZF+VErNgfAYTBgTXwNEAUNhrGCHBKEZAADgADI16RBBIdphQBvTeFxbkjGgBHfHwdiUfooBjESJQOBweRq2lsOBSzBar7iH5IP7ExqCESYTbbVHELG46VIo0oxowKnMxIEZBpZ3TTC6LEyESGkC8ACstIEcAAKlT5caASBbjshFB1bBFmAGVQAIIpxYwZz6ImCDweIA).

Thomaash commented on lib/network/modules/components/nodes/shapes/Image.js:
```diff
@@ -108,7 +115,20 @@ class Image extends CircleImageBase {
     this.resize();
-    this._updateBoundingBox(x, y);
+
+    if(this.options.shapeProperties.useImageCoords){
+      this.left = x;
+      this.top = y;
+    }
+    else{
+      this.left = x - this.width / 2;
+      this.top = y - this.height / 2;
+    }
+
+    this.boundingBox.left   = this.left;
+    this.boundingBox.top    = this.top;
+    this.boundingBox.bottom = this.top + this.height;
+    this.boundingBox.right  = this.left + this.width;
```
dtto

cristianoccazinsp commented:
Thanks for the feedback. I will see about resolving the test issues.

About why only for images.  You are right, it could be technically be added to all other shapes too, but I don't know if they would be useful at all, mostly because shapes are quite limited (no custom width/height for rectangles nor other shapes, just a generic `size` prop).  For my use case, it was really important that SVG (rect) images get rendered at the (0, 0) position instead of width/2, height/2 because of some math issues.  I also see there are some other props that are "image only", so I see no problems with adding an extra one.

For the label issues, I couldn't seem to find where to adjust the Label top/left positions when this new prop is used. That's something I need to still play around with, but I needed this change asap for my project.

cristianoccazinsp commented:
Added a commit to address the above.

The prop can be now used with image and circularImage.  I'm afraid that there would be a lot of code duplication if we add it to the other shapes (just like there are lots of width/2 and height/2 calls in every shape).

Can you reconsider having this new setting just for the image and circularImage shapes?  Truth is, it is really not useful except on a very unusual case (like my own).

Here are the two new samples with the new prop applied:

<img width="764" alt="Screen Shot 2020-06-18 at 18 08 56" src="https://user-images.githubusercontent.com/48869228/85073707-a41d9d80-b191-11ea-84e7-550386c207ef.png">
<img width="626" alt="Screen Shot 2020-06-18 at 18 09 07" src="https://user-images.githubusercontent.com/48869228/85073711-a7188e00-b191-11ea-97ba-de9f952b26e6.png">

cristianoccazinsp commented:
Getting some time out on a test that took over 4000ms, but my mac is also quite slow due to all the multiple open editors and tools. Not sure if that's the problem.

cristianoccazinsp commented:
Is the test still timing out, or just giving an error?  I cannot seem to view the test results without making a circleci account (which I won't due to private repo access).

Thomaash commented:
> Is the test still timing out, or just giving an error? I cannot seem to view the test results without making a circleci account (which I won't due to private repo access).

I increased the timeouts in master. If you update your branch it should work better.

cristianoccazinsp commented:
Test took a lot of time and is still failing, can't figure out what's going on :D

Thomaash commented on docs/network/nodes.html:
```diff
@@ -1071,6 +1072,14 @@ <h3>Options</h3>
+            <td class="indent">shapeProperties.useImageCoords</td>
+            <td>Boolean</td>
+            <td><code>false</code></td>
+            <td>This property only applies to the <code>image</code> and <code>circularImage</code> shapes.
+                When true, the image's (0,0) coords are used for drawing.
```
I have no idea what this means. Used for drawing? How? Why? What happens if it's not true?

cristianoccazinsp replied:
I'm really not sure how to describe the drawing process.  Currently, when an image/svg is used, the (x, y) of the node will be the center of the image (using width and height / 2). When this prop is set to true, the behaviour changes so the (x, y) of the node matches to the (0, 0) of the image.  Can you propose a concise description for this behaviour in a few words?

Thomaash replied:
I suggest some polished version of:
> When an image/svg is used, the (x, y) of the node will be the center of the image (using width and height / 2). When this prop is set to true, the behaviour changes so the (x, y) of the node matches to the (0, 0) of the image.

Thomaash commented on examples/network/nodeStyles/circularImagesWithPosition.html:
```diff
@@ -0,0 +1,109 @@
+<!doctype html>
+<html>
+<head>
+  <title>Vis Network | Node Styles | Circular Images</title>
```
Could you consolidate those two examples into a single one? Also change the title (that's the thing that actually appears in examples index).

cristianoccazinsp replied:
I agree with the title change. But won't it get too messy to have both examples in the same file?  I'm just following what's already done, and there seems to be 1 file for each image and circular image case.

Thomaash replied:
It's a single example, that's why I want you to merge it. You are showcasing single thing: positioning. It works with two node shapes so just make some nodes image and some circular image.

Thomaash commented on examples/network/nodeStyles/images.html:
```diff
@@ -68,7 +68,7 @@
     }
   </script>
-  
+
```
Why?

cristianoccazinsp replied:
That's just visual studio code removing trailing spaces I believe.

Thomaash commented on lib/network/modules/components/nodes/shapes/CircularImage.js:
```diff
@@ -58,11 +58,22 @@ class CircularImage extends CircleImageBase {
-    this.left = x - this.width / 2;
-    this.top = y - this.height / 2;
+
+    var labelX = x,
+      labelY = y;
```
Don't use var, please. Use const (or let if you really have to reassign the value).

Thomaash commented on lib/network/modules/components/nodes/shapes/Image.js:
```diff
@@ -58,8 +58,19 @@ class Image extends CircleImageBase {
-    this.left = x - this.width / 2;
-    this.top = y - this.height / 2;
+
+    var labelX = x,
+      labelY = y;
```
Are the changes in this file any different to the changes in CircularImage.js?

cristianoccazinsp replied:
Image overrides `draw` from CircularImage to add some extra code (and pretty much copy the rest of the existing code). I'm just mimicking what's already done.  I don't think it would be worth it to add yet another function call for just these few lines.

Thomaash replied:
The code quality here is not particularly good but I understand if you don't want to dive too deep into this.

Thomaash commented on lib/network/modules/components/nodes/util/CircleImageBase.js:
```diff
@@ -47,7 +47,7 @@ class CircleImageBase extends NodeBase {
    *
    * The images can be updated after the initial setting of options;
-   * therefore, this method needs to be reentrant. 
+   * therefore, this method needs to be reentrant.
```
Please, don't reformat files, especially don't if you don't change anything in it.

cristianoccazinsp replied:
Visual Studio code removing trailing spaces...

Thomaash replied:
I don't care what tool you use to remove them, just don't. The problem is that you'll be credited as the author of these lines and when I need to find out why was something introduced your name and commit will pop up simply because you didn't turn off reformatting in VS Code.

Thomaash commented on lib/network/options.js:
```diff
@@ -403,6 +403,7 @@ let allOptions = {
       interpolation: { boolean: bool },
       useImageSize: { boolean: bool },
       useBorderWithImage: { boolean: bool },
+      useImageCoords: { boolean: bool },
```
This is not very descriptive name. I suggest (this is up for discussion of course):
```typescript
interface shapeProperties {
  coordinateOrigin?:
    | "center" // default
    | "top-left";
}
```
This way people would have better chance at intuitively guess what it does.

cristianoccazinsp replied:
This makes sense.

cristianoccazinsp commented:
Pushed a new commit to address the following:

- Renamed `useImageCoords`
- Updated docs to better word these
- Code style fixes
- Merged the example into one

Please disregard any trailing space removed by visual studio code. Looks like someone left a bunch of those and each time a file is saved they get removed!

Thomaash commented on examples/network/nodeStyles/imagesWithCoordinateOrigin.html:
```diff
@@ -0,0 +1,91 @@
+            useImageSize: true,
+            useBorderWithImage: false,
+            interpolation: false,
+            coordinateOrigin: 'top-left'
```
It would be great to also showcase `"center"` so that people can see them side by side for comparison.

cristianoccazinsp replied:
Thanks for the feedback. Didn't think those formatting "fixes" were such an issue but it makes sense now.  I agree with the other changes too, I will see to add those in shortly during the weekend.

cristianoccazinsp replied:
Can't really find out what validation you're talking about with `add input validation to Node.parseOptions in vis-network/lib/network/modules/components/Node.js`. How would the validation look like? Just check for the `center` and `top-left` values? I can't understand what's new options and parent options, and values are often undefined (not loaded a default value here)

cristianoccazinsp commented:
Tried to merge from master to remove the unwanted formatting fixes, and looks like it brought up a bunch of other commits/changes.

cristianoccazinsp commented:
Hmm, I havent had this problem before when rebasing like this. Is it
fixable now with those commands?

El dom., 5 de julio de 2020 16:38, Tomáš Vyčítal <notifications@github.com>
escribió:

> *@Thomaash* requested changes on this pull request.
>
> This is a known (by design) bug in GitHub, it's insane to review something
> like this.
>
> You should rebase (git pull --rebase upstream master assuming upstream
> points to the visjs repo) and force push instead of merging.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/visjs/vis-network/pull/812#pullrequestreview-442701370>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALU263GOAKHA7AP5QQJQW5DR2DJDXANCNFSM4OBYB6RQ>
> .
>

Thomaash commented:
> Hmm, I havent had this problem before when rebasing like this. Is it fixable now with those commands? El dom., 5 de julio de 2020 16:38, Tomáš Vyčítal <notifications@github.com> escribió:
> […](#)
> *@Thomaash* requested changes on this pull request. This is a known (by design) bug in GitHub, it's insane to review something like this. You should rebase (git pull --rebase upstream master assuming upstream points to the visjs repo) and force push instead of merging. — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <[#812 (review)](https://github.com/visjs/vis-network/pull/812#pullrequestreview-442701370)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ALU263GOAKHA7AP5QQJQW5DR2DJDXANCNFSM4OBYB6RQ> .

Yeah, it should be. If not, you can just clone the upstream master again, copy the files over, commit them (we'll squash the history anyway) and force push.

cristianoccazinsp commented:
I think I'm going to just make a fresh PR, this branch is all screwed up, I've never seen github act like this when pulling down commits from master.

cristianoccazinsp commented:
New PR here https://github.com/visjs/vis-network/pull/844: also fixed some extra files I had missed when reverting the auto style fixes from VS Code
